alf/algorithms/agent_helpers.py

```diff
@@ -64,61 +65,81 @@ def accumulate_algortihm_rewards(rewards, weights, names, summary_prefix,
             `training_info`.
+        training_info (nested Tensor): information collected for training
+            algorithms. It is batched from each `info` returned by `train_step()`.
+
+    Returns:
+        loss_info (LossInfo): the accumulated loss info
     """
 
-    def _make_rl_training_info(training_info, name):
-        if training_info.rollout_info == ():
-            rollout_info = ()
+    def _update_loss(loss_info, training_info, algorithm, name):
+        if isinstance(algorithm, RLAlgorithm):
+            new_loss_info = algorithm.calc_loss(
+                _make_rl_training_info(training_info, name))
         else:
-            rollout_info = getattr(training_info.rollout_info, name)
-        info = getattr(training_info.info, name)
-        return training_info._replace(info=info, rollout_info=rollout_info)
-
-    def _update_loss(loss_info, training_info, name, algorithm):
-        if algorithm is None:
-            return loss_info
-        new_loss_info = algorithm.calc_loss(getattr(training_info.info, name))
-        loss_info.extra[name] = new_loss_info.extra
-        return LossInfo(
-            loss=add_ignore_empty(loss_info.loss, new_loss_info.loss),
-            scalar_loss=add_ignore_empty(loss_info.scalar_loss,
-                                         new_loss_info.scalar_loss),
-            extra=loss_info.extra)
-
-    assert len(algorithms) > 0
-    assert len(algorithms) == len(names)
-    assert isinstance(algorithms[0], RLAlgorithm)
+            new_loss_info = algorithm.calc_loss(
+                getattr(training_info.info, name))
+        if loss_info is None:
+            return new_loss_info._replace(extra=dict(name=new_loss_info.extra))
+        else:
+            loss_info.extra[name] = new_loss_info.extra
+            return LossInfo(
+                loss=add_ignore_empty(loss_info.loss, new_loss_info.loss),
+                scalar_loss=add_ignore_empty(loss_info.scalar_loss,
+                                             new_loss_info.scalar_loss),
+                extra=loss_info.extra)
 
-    rl_loss_info = algorithms[0].calc_loss(
-        _make_rl_training_info(training_info, names[0]))
-    loss_info = rl_loss_info._replace(extra=dict(rl=rl_loss_info.extra))
-    for alg, name in zip(algorithms[1:], names[1:]):
-        loss_info = _update_loss(loss_info, training_info, name, alg)
+    assert len(algorithms) == len(names)
+    loss_info = None
+    for alg, name in zip(algorithms, names):
+        loss_info = _update_loss(loss_info, training_info, alg, name)
+    assert loss_info is not None, "No loss info is calculated!"
     return loss_info
+
+
+def after_update(algorithms, names, training_info):
+    """For each provided algorithm, call its `after_update()` to do things after
+    the agent completes one gradient update (i.e. `update_with_gradient()`).
+
+    Args:
+        algorithms (list[Algorithm]): the list of algorithms whose `after_update`
```

Comment:
list[RLAlgorithm]. Or update the code to differentiate RLAlgorithm with Algorithm

Reply:
I think in general this API can be open to any Algorithm. For example, after each update I want to increase a counter for a non-RL algorithm. I've updated the code to handle differently between the two.